### PR TITLE
Dev/gfql serialization

### PR DESCRIPTION
**Pull request by lmeyerov**
Adding serialization of GFQL queries in preparation for exposing as a public FEP endpoint on datasets

```python
pattern = [n(), e(), n()]
o = graphistry.compute.chain.to_json(pattern)
pattern2 = graphistry.compute.chain.from_json(o)
g.chain(pattern2).plot()
```

---

### Added

* GFQL `Chain` AST object
* GFQL query serialization - `Chain`, `ASTObject`, and `ASTPredict` implement `ASTSerializable`
  - Ex:`Chain.from_json(Chain([n(), e(), n()]).to_json())`
* GFQL predicate `is_year_end`

### Docs

* GFQL in readme.md

### Changes

* Refactor `ASTEdge`, `ASTNode` field naming convention to match other `ASTSerializable`s (`e.edge_query` vs `e._edge_query`)

### Breaking 🔥

* GFQL `e()` now aliases `e_undirected` instead of the base class `ASTEdge`

### CODE REVIEW

File: graphistry/compute/predicates/numeric.py [hunk 1]
```diff
@@ -1,80 +1,162 @@
...
     def __init__(self, val: float) -> None:
         self.val = val
 
     def __call__(self, s: pd.Series) -> pd.Series:
         return s == self.val
 
+    def to_json(self, validate=True) -> dict:
+        if validate:
+            self.validate()
+        return {'type': 'EQ', 'val': self.val}
```
Comment (aucahuasi) [inline graphistry/compute/predicates/numeric.py:126]: There are many calls with this pattern at the end:
```python
return {'type': 'EQ', 'val': self.val}
```
I wonder if we can add some util method in the base class NumericASTPredicate to create that map using self.val internally, so we would have something like this:
```python
return self._to_json('EQ')
```
and the implementation in the base would be:
```python
def _to_json(self, op:string):
  return {'type': op, 'val': self.val}
```

**Review by aucahuasi — APPROVED** [review]
LGTM!

File: graphistry/compute/predicates/numeric.py (hunk 1 above)
Comment (lmeyerov) [reply graphistry/compute/predicates/numeric.py:126]: ah good idea

like auto-mixin `self.__class__.__name__`

Comment (aucahuasi) [reply graphistry/compute/predicates/numeric.py:126]: Yes it can be a mixing as well.

Comment (lmeyerov) [reply graphistry/compute/predicates/numeric.py]: thanks, i was able to cut like 300+ loc this way

**Pull request merged**